lib/c_glib/src/thrift/c_glib/transport/thrift_ssl_socket.c

```diff
@@ -0,0 +1,700 @@
+			socklen_t saLength = sizeof(struct sockaddr_storage);
+
+			if (getpeername(socket->sd, (struct sockaddr*)&sa, &saLength) != 0) {
+				sa.ss_family = AF_UNSPEC;
+			}
+			authorization_result = cls->authorize_peer(transport, cert, &sa, error);
+		}
+		if(cert!=NULL){
+			X509_free(cert);
+		}
+	}
+
+	return authorization_result;
+}
+
+
+/* initializes the instance */
+static void
+thrift_ssl_socket_init (ThriftSSLSocket *socket)
+{
+	socket->ssl = NULL;
+	socket->ctx = NULL;
+	socket->server = FALSE;
+}
+
+/* destructor */
+static void
+thrift_ssl_socket_finalize (GObject *object)
+{
+	ThriftSSLSocket *socket = THRIFT_SSL_SOCKET (object);
+	GError *error=NULL;
+	if(socket->ssl != NULL)
+	{
+		thrift_ssl_socket_close(THRIFT_TRANSPORT(object), &error);
+		socket->ssl=NULL;
+		socket->ctx=NULL;
+	}
+}
+
+/* property accessor */
+void
+thrift_ssl_socket_get_property (GObject *object, guint property_id,
+		GValue *value, GParamSpec *pspec)
+{
+	ThriftSSLSocket *socket = THRIFT_SSL_SOCKET (object);
+
+	THRIFT_UNUSED_VAR (pspec);
+	g_test_message("Getting property id %d", value);
+
+	switch (property_id)
+	{
+	case PROP_THRIFT_SSL_SOCKET_CONTEXT:
+		g_value_set_pointer (value, socket->ctx);
+		break;
+	}
+}
+
+/* property mutator */
+void
+thrift_ssl_socket_set_property (GObject *object, guint property_id,
+		const GValue *value, GParamSpec *pspec)
+{
+	//ThriftSocketClass *tsc= THRIFT_SOCKET_CLASS(object);
+	ThriftSSLSocket *socket = THRIFT_SSL_SOCKET (object);
+
+	THRIFT_UNUSED_VAR (pspec);
+	g_test_message("Setting property id %d", property_id);
+	switch (property_id)
+	{
+	case PROP_THRIFT_SSL_SOCKET_CONTEXT:
+		socket->ctx = g_value_get_pointer(value); // We copy the context
+		break;
+		//	default:
+		//		thrift_socket_set_property(object, property_id, value, pspec);
+		//		break;
+	}
+}
+
+void
+thrift_ssl_socket_initialize_openssl(void)
+{
+	if(thrift_ssl_socket_openssl_initialized){
+		return;
+	}
+	thrift_ssl_socket_openssl_initialized=TRUE;
+	SSL_library_init();
+	ERR_load_crypto_strings();
+	SSL_load_error_strings();
+	ERR_load_BIO_strings();
+
+#if (OPENSSL_VERSION_NUMBER < OPENSSL_VERSION_NO_THREAD_ID)
+	CRYPTO_set_id_callback(callbackThreadID);
+#endif
+	CRYPTO_set_locking_callback(thrift_ssl_socket_static_locking_callback);
+	// dynamic locking
+	CRYPTO_set_dynlock_create_callback(thrift_ssl_socket_dyn_lock_create_callback);
+	CRYPTO_set_dynlock_lock_callback(thrift_ssl_socket_dyn_lock_callback);
+	CRYPTO_set_dynlock_destroy_callback(thrift_ssl_socket_dyn_lock_destroy_callback);
+}
+
+
+void thrift_ssl_socket_finalize_openssl(void)
+{
+
+	// FIXME This should not be here
+	if (thrift_ssl_socket_global_context != NULL) {
+		SSL_CTX_free(thrift_ssl_socket_global_context);
+		thrift_ssl_socket_global_context = NULL;
+	}
+
+	if (!thrift_ssl_socket_openssl_initialized) {
+		return;
+	}
+	thrift_ssl_socket_openssl_initialized = FALSE;
+#if (OPENSSL_VERSION_NUMBER < OPENSSL_VERSION_NO_THREAD_ID)
+	CRYPTO_set_id_callback(NULL);
+#endif
```

Comment:
It's original cpp lib TSSLSocket problem but we're supposed to call CRYPTO_THREADID_set_callback for `#else` path.
It does not seem to cause any problem on most platforms, according to the doc though.